File: app/src/main/java/com/example/zipfileinmemoryjni/ByteBufferChannel.java
package com.example.zipfileinmemoryjni;

import org.apache.commons.compress.utils.SeekableInMemoryByteChannel;

import java.nio.ByteBuffer;

import static java.lang.Math.min;

public final class ByteBufferChannel extends SeekableInMemoryByteChannel {
    private final ByteBuffer buf;

    public ByteBufferChannel(ByteBuffer buffer) {
        if (buffer == null) throw new NullPointerException();
        buf = buffer;
    }

    @Override
    public synchronized int read(ByteBuffer dst) {
        if (buf.remaining() == 0) return -1;

        int count = min(dst.remaining(), buf.remaining());
        if (count > 0) {
            ByteBuffer tmp = buf.slice();
            tmp.limit(count);
            dst.put(tmp);
            buf.position(buf.position() + count);
        }
        return count;
    }

    @Override
    public synchronized long position() {
        return buf.position();
    }

    @Override
    public synchronized ByteBufferChannel position(long newPosition) {
        if ((newPosition | Integer.MAX_VALUE - newPosition) < 0)
            throw new IllegalArgumentException();
        buf.position((int) newPosition);
        return this;
    }

    @Override
    public synchronized long size() {
        return buf.limit();
    }

    @Override
    public boolean isOpen() {
        return true;
    }

    @Override
    public void close() {
    }
}

File: app/src/main/java/com/example/zipfileinmemoryjni/ByteBufferChannel.kt
package com.example.zipfileinmemoryjni

// From https://gist.github.com/ncruces/0e6e625899645f5a88e4d7444d696130 and converted to Kotlin
// with some minor modifications.

import java.nio.ByteBuffer
import java.nio.channels.SeekableByteChannel

class ByteBufferChannel(buffer: ByteBuffer) : SeekableByteChannel {
    private val buf: ByteBuffer = buffer

    @Synchronized
    override fun read(dst: ByteBuffer): Int {
        if (buf.remaining() == 0) return -1
        val count = dst.remaining().coerceAtMost(buf.remaining())
        if (count > 0) {
            buf.slice().apply {
                limit(count)
                dst.put(this)
            }
            buf.position(buf.position() + count)
        }
        return count
    }

    @Synchronized
    override fun position(newPosition: Long): ByteBufferChannel {
        require(newPosition or Int.MAX_VALUE - newPosition >= 0)
        buf.position(newPosition.toInt())
        return this
    }

    override fun write(src: ByteBuffer?): Int {
        TODO("Not yet implemented")
    }

    @Synchronized
    override fun size() = buf.limit().toLong()

    @Synchronized
    override fun position() = buf.position().toLong()

    override fun isOpen() = true

    override fun close() {}

    override fun truncate(size: Long): SeekableByteChannel {
        TODO("Not yet implemented")
    }
}